master/internal/stream/system.go

```diff
@@ -167,6 +239,26 @@ func (ps PublisherSet) Websocket(socket *websocket.Conn, c echo.Context) error {
 		streamer.Close()
 	}()
 
+	// detect & handle permission change
+	go func() {
+		if err := AuthZProvider.Get().WaitForPermissionChange(); err != nil {
```

Comment:
The fact that this is a blocking call indicates that you'll be leaking goroutines.

Imagine a thousand streamers connecting, with an average lifetime of 10 minutes.

Suppose a cluster goes a day without rebooting.

That's 150 thousand goroutines that are all going to be blocking on this call.  There's no mechanism for them to get preempted when their associated websocket connections break.

This is another reason why I think using WaitForPermissionsChange() is a lot harder than just having the NOTIFY listener call ps.Reset() directly, because for WaitForPermissionsChange to work you also need to add preemption to this blocking call.

Reply:
The perk of this approach is so that we don't have to include any references to implementation details for watching for permission changes inside of `system.go`, especially since in OSS permission changes don't occur since users have global view permissions.

I do agree, that leaking goroutines is not the right call here, especially to this degree, if we were to move this into NewPublisherSet, that would limit it to 1 go-routine (dedicated to permission changes) per PublisherSet , rather than per websocket. If we were to set it up this way, and have the NOTIFY listener call ps.Restart() in `NewPublisherSet`, or even keep WaitForPermissionChange(), we'd still be running into the issue where we'd be stuck on a blocking call `<-permissionListener.Notify`. So setting up setting up preempt-ability is a must in both cases.   

We could set it up inside a loop w/ a select that is watching for:
- time out then ping 
- if the publisher set closed
- or a permission change to occur

Similar to what's being done for already on the other listener.

I like the idea of OSS getting to ignoring this step completely since in OSS appearance/disappearance can't occur. I could change `WaitForPermissonChange` to `GetWaitForPermissionChangeFunc() func() error {}` or even `GetPermissionChangeListener() (*pq.Listener, error)`. That way we can still handle the difference in behavior the authz call is synchronous & the result can be used asynchronously.

Reply:
Instead of having a blocking call in authz, authz will a way to get the *listener, and then handling it accordingly in `system.go`